ERCS/erc-7590.md

```diff
@@ -0,0 +1,178 @@
+}
+```
+
+
+## Rationale
+
+### Pull Mechanism
+
+We propose using a pull mechanism, where the contract transfers the token to itself, instead of receiving it via "safe transfer" for 2 reasons:
+
+1. Customizability with Hooks. By initiating the process this way, smart contract developers have the flexibility to execute specific actions before and after transferring the tokens.
+
+2. Lack of transfer with callback: [ERC-20](./eip-20.md) tokens lack a standardized transfer with callback method, such as the "safeTransfer" on [ERC-721](./eip-721.md), which means there is no reliable way to notify the receiver of a successful transfer, nor to know which is the destination token is.
+
+This has the disadvantage of requiring approval of the token to be transferred before actually transferring it into an NFT.
+
+### Granular vs Generic
+
+We considered 2 ways of presenting the proposal:
+1. A granular approach where there is an independent interface for each type of held token.
+2. A universal token holder which could also hold and transfer [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md).
+
+An implementation of the granular version is slightly cheaper in gas, and if you're using just one or two types, it's smaller in contract size. The generic version is smaller and has single methods to send or receive, but it also adds some complexity by always requiring Id and amount on transfer methods. Id not being necessary for [ERC-20](./eip-20.md) and amount not being necessary for [ERC-721](./eip-721.md).
+
+We also considered that due to the existence of safe transfer methods on both [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md), and the commonly used interfaces of `IERC721Receiver` and `IERC1155Receiver`, there is not much need to declare an additional interface to manage such tokens. However, this is not the case for [ERC-20](./eip-20.md), which does not include a method with a callback to notify the receiver of the transfer.
+
+For the aforementioned reasons, we decided to go with a granular approach.
+
+
+## Backwards Compatibility
+
+No backward compatibility issues found.
+
+## Test Cases
+
+Will be added.
+
+## Reference Implementation
+
+Will be added.
+
+## Security Considerations
+
+The same security considerations as with [ERC-721](./eip-721.md) apply: hidden logic may be present in any of the functions, including burn, add resource, accept resource, and more.
+
+Caution is advised when dealing with non-audited contracts.
+
+Implementations MUST use the message sender as from parameter when they are transferring tokens into an NFT. Otherwise, since the current contract needs approval, it could potentially pull the external tokens into a different NFT.
+
+To prevent a seller from front running the sale of an NFT holding [ERC-20](./eip-20.md) tokens to transfer out such tokens before a sale is executed, marketplaces must beware of the `erc20TransferOutNonce` and revert if it has changed since listed.
+
```

Comment:
I'd also mention that if ERC-20 tokens are transferred directly to the NFT contract, they are lost.